File: Helper/Core/Scroll/ScrollUtility.m
//
// --------------------------------------------------------------------------
// ScrollUtility.m
// Created for Mac Mouse Fix (https://github.com/noah-nuebling/mac-mouse-fix)
// Created by Noah Nuebling in 2019
// Licensed under the MMF License (https://github.com/noah-nuebling/mac-mouse-fix/blob/master/License)
// --------------------------------------------------------------------------
//

#import "ScrollUtility.h"
#import "Scroll.h"
#import <Cocoa/Cocoa.h>
#import "IOHIDEventTypes.h"
#import "SharedUtility.h"
#import "ModificationUtility.h"
#import "Logging.h"

@implementation ScrollUtility


static NSDictionary *_MFScrollPhaseToIOHIDEventPhase;
+ (NSDictionary *)MFScrollPhaseToIOHIDEventPhase {
    return _MFScrollPhaseToIOHIDEventPhase;
}

+ (void)load {
    _MFScrollPhaseToIOHIDEventPhase = @{
        @(kMFPhaseNone):       @(kIOHIDEventPhaseUndefined),
        @(kMFPhaseStart):      @(kIOHIDEventPhaseBegan),
        @(kMFPhaseLinear):     @(kIOHIDEventPhaseChanged),
        @(kMFPhaseMomentum):   @(kIOHIDEventPhaseChanged),
        @(kMFPhaseEnd):        @(kIOHIDEventPhaseEnded),
    };
}

+ (CGEventRef)createPixelBasedScrollEventWithValuesFromEvent:(CGEventRef)event {
    
    /// Basically creates a copy of a scroll event.
    /// \discussion
    /// When multithreading from within `ScrollControl -> eventTapCallback()` events would become invalid and unusable in the new thread.
    /// Using CGEventCreateCopy didn't help, but this does fix the issue. Not sure why.
    /// Xcode Analysis warns of potential memory leak here, even though we have `create` in the name. Maybe we should flag the return with `CF_RETURNS_RETAINED`
    /// This doesn't produce identically behaving events (See https://github.com/noah-nuebling/mac-mouse-fix/issues/61)
    ///     So we made a more general version of this function, which copies over _all_ fields, at `Utility_HelperApp:createEventWithValuesFromEvent:` ... which couldn't produce identical results either... so we're back to trying to make CGEventCreateCopy work.
    
    CGEventRef newEvent = CGEventCreateScrollWheelEvent(NULL, kCGScrollEventUnitPixel, 1, 0);
    NSArray *valueFields = @[@(kCGScrollWheelEventDeltaAxis1),
                             @(kCGScrollWheelEventDeltaAxis2),
                             @(kCGScrollWheelEventPointDeltaAxis1),
                             @(kCGScrollWheelEventPointDeltaAxis2),
                             @(kCGScrollWheelEventFixedPtDeltaAxis1),
                             @(kCGScrollWheelEventFixedPtDeltaAxis2)];
    for (NSNumber *f in valueFields) {
        int64_t ogVal = CGEventGetIntegerValueField(event, f.intValue);
        CGEventSetIntegerValueField(newEvent, f.intValue, ogVal);
    }
    return newEvent;
}

+ (CGEventRef)createNormalizedEventWithPixelValue:(int)lineHeight {
    
    /// Creates a vertical scroll event with a line delta value of 1 and a pixel value of `lineHeight`
    /// \discussion Xcode Analysis warns of potential memory leak here, even though we have `create` in the name. Maybe we should flag the return with `CF_RETURNS_RETAINED`
    
    /// invert vertical
    CGEventRef event = CGEventCreateScrollWheelEvent(NULL, kCGScrollEventUnitPixel, 1, 0);
    CGEventSetIntegerValueField(event, kCGScrollWheelEventDeltaAxis1, 1);
    CGEventSetIntegerValueField(event, kCGScrollWheelEventPointDeltaAxis1, lineHeight);
    CGEventSetDoubleValueField(event, kCGScrollWheelEventFixedPtDeltaAxis1, lineHeight);
    
    DDLogInfo(@"Normalized scroll event values:");
    DDLogInfo(@"%lld",CGEventGetIntegerValueField(event, kCGScrollWheelEventDeltaAxis1));
    DDLogInfo(@"%lld",CGEventGetIntegerValueField(event, kCGScrollWheelEventPointDeltaAxis1));
    DDLogInfo(@"%f",CGEventGetDoubleValueField(event, kCGScrollWheelEventFixedPtDeltaAxis1));
    
    return event;
}
/// Inverts the diection of a given scroll event if dir is -1.
/// @param event Event to be inverted
/// @param dir Either 1 or -1. 1 Will leave the event unchanged.
+ (CGEventRef)invertScrollEvent:(CGEventRef)event direction:(int)dir {
    // invert vertical
    long long line1 = CGEventGetIntegerValueField(event, kCGScrollWheelEventDeltaAxis1);
    long long point1 = CGEventGetIntegerValueField(event, kCGScrollWheelEventPointDeltaAxis1);
    long long fixedPt1 = CGEventGetIntegerValueField(event, kCGScrollWheelEventFixedPtDeltaAxis1);
    CGEventSetIntegerValueField(event, kCGScrollWheelEventDeltaAxis1, line1 * dir);
    CGEventSetIntegerValueField(event, kCGScrollWheelEventPointDeltaAxis1, point1 * dir);
    CGEventSetIntegerValueField(event, kCGScrollWheelEventFixedPtDeltaAxis1, fixedPt1 * dir);
    // invert horizontal
    long long line2 = CGEventGetIntegerValueField(event, kCGScrollWheelEventDeltaAxis2);
    long long point2 = CGEventGetIntegerValueField(event, kCGScrollWheelEventPointDeltaAxis2);
    long long fixedPt2 = CGEventGetIntegerValueField(event, kCGScrollWheelEventFixedPtDeltaAxis2);
    CGEventSetIntegerValueField(event, kCGScrollWheelEventDeltaAxis2, line2 * dir);
    CGEventSetIntegerValueField(event, kCGScrollWheelEventPointDeltaAxis2, point2 * dir);
    CGEventSetIntegerValueField(event, kCGScrollWheelEventFixedPtDeltaAxis2, fixedPt2 * dir);
    return event;
}
+ (CGEventRef)makeScrollEventHorizontal:(CGEventRef)event {
    
    long long line1 = CGEventGetIntegerValueField(event, kCGScrollWheelEventDeltaAxis1);
    long long point1 = CGEventGetIntegerValueField(event, kCGScrollWheelEventPointDeltaAxis1);
    long long fixedPt1 = CGEventGetIntegerValueField(event, kCGScrollWheelEventFixedPtDeltaAxis1);
    
    CGEventSetIntegerValueField(event, kCGScrollWheelEventDeltaAxis1, 0);
    CGEventSetIntegerValueField(event, kCGScrollWheelEventPointDeltaAxis1, 0);
    CGEventSetIntegerValueField(event, kCGScrollWheelEventFixedPtDeltaAxis1, 0);
    
    CGEventSetIntegerValueField(event, kCGScrollWheelEventDeltaAxis2, line1);
    CGEventSetIntegerValueField(event, kCGScrollWheelEventPointDeltaAxis2, point1);
    CGEventSetIntegerValueField(event, kCGScrollWheelEventFixedPtDeltaAxis2, fixedPt1);
    
    return event;
}

+ (void)logScrollEvent:(CGEventRef)event {
    
    long long line1 = CGEventGetIntegerValueField(event, kCGScrollWheelEventDeltaAxis1);
    long long point1 = CGEventGetIntegerValueField(event, kCGScrollWheelEventPointDeltaAxis1);
    long long fixedPt1 = CGEventGetIntegerValueField(event, kCGScrollWheelEventFixedPtDeltaAxis1);
    
    long long line2 = CGEventGetIntegerValueField(event, kCGScrollWheelEventDeltaAxis2);
    long long point2 = CGEventGetIntegerValueField(event, kCGScrollWheelEventPointDeltaAxis2);
    long long fixedPt2 = CGEventGetIntegerValueField(event, kCGScrollWheelEventFixedPtDeltaAxis2);

    
    DDLogInfo(@"Axis 1:");
    DDLogInfo(@"  Line: %lld", line1);
    DDLogInfo(@"  Point: %lld", point1);
    DDLogInfo(@"  FixedPt: %lld", fixedPt1);
    
    DDLogInfo(@"Axis 2:");
    DDLogInfo(@"  Line: %lld", line2);
    DDLogInfo(@"  Point: %lld", point2);
    DDLogInfo(@"  FixedPt: %lld", fixedPt2);
    
}

+ (BOOL)point:(CGPoint)p1 isAboutTheSameAs:(CGPoint)p2 threshold:(int)th {
    if (abs((int)(p2.x - p1.x)) > th || abs((int)(p2.y - p1.y)) > th) {
        return NO;
    }
    return YES;
}

/// \note 0 is considered both positive and negative
+ (BOOL)sameSign:(double)n and:(double)m {
    if (n == 0 || m == 0) {
        return true;
    }
    if ([SharedUtility signOf:n] == [SharedUtility signOf:m]) {
        return true;
    }
    return false;
}

+ (MFAxis)axisForVerticalDelta:(int64_t)deltaV horizontalDelta:(int64_t)deltaH {
    
    NSCAssert(deltaV == 0 || deltaH == 0, @"Scroll event is not parallel to an axis.");
    
    MFAxis axis = kMFAxisVertical;
    if (deltaH != 0) {
        axis = kMFAxisHorizontal;
    }
    
    return axis;
}

+ (MFDirection)directionForInputAxis:(MFAxis)axis
                           inputDelta:(int64_t)inputDelta
                        invertSetting:(MFScrollInversion)invertSetting
                   horizontalModifier:(BOOL)horizontalModifier {
    
    /// Axis direction
    
    MFScrollInversion inputAxisDirection = [SharedUtility signOf:inputDelta];
    MFScrollInversion effectiveAxisDirection = inputAxisDirection * invertSetting;
    
    /// Direction
    
    if (axis == kMFAxisHorizontal || horizontalModifier) {
        /// H
        if (effectiveAxisDirection == -1) {
            return kMFDirectionLeft;
        } else {
            return kMFDirectionRight;
        }
    } else {
        /// V
        if (effectiveAxisDirection == -1) {
            return kMFDirectionDown;
        } else {
            return kMFDirectionUp;
        }
    }
}


static BOOL _mouseDidMove = NO;
+ (BOOL)mouseDidMove {
    return _mouseDidMove;
}
static CGPoint _previousMouseLocation;
/// Checks if cursor did move since the last time this function was called. Writes result into `_mouseDidMove`.
///     Storing result in _mouseDidMove instead of returning so that different parts of the program can reuse this info
/// Passing in event for optimization. Not sure if significatnt.
+ (void)updateMouseDidMoveWithEvent:(CGEventRef)event {
    CGPoint mouseLocation = CGEventGetLocation(event);
    _mouseDidMove = ![ScrollUtility point:mouseLocation
                          isAboutTheSameAs:_previousMouseLocation
                                 threshold:10];
    _previousMouseLocation = mouseLocation;
}

static BOOL _frontMostAppDidChange;
+ (BOOL)frontMostAppDidChange {
    return _frontMostAppDidChange;
}
static NSRunningApplication *_previousFrontMostApp;
/// Checks if frontmost application changed since the last time this function was called. Writes result into `_frontMostAppDidChange`.
+ (void)updateFrontMostAppDidChange {
    NSRunningApplication *frontMostApp = NSWorkspace.sharedWorkspace.frontmostApplication;
    _frontMostAppDidChange = ![frontMostApp isEqual:_previousFrontMostApp];
    _previousFrontMostApp = frontMostApp;
}

@end

File: Helper/Core/Scroll/ScrollUtility.swift
//
// --------------------------------------------------------------------------
// ScrollUtility.swift
// Created for Mac Mouse Fix (https://github.com/noah-nuebling/mac-mouse-fix)
// Created by Noah Nuebling in 2019
// Licensed under the MMF License (https://github.com/noah-nuebling/mac-mouse-fix/blob/master/License)
// --------------------------------------------------------------------------
//

#if IS_HELPER

import Cocoa
import IOKit

private let kCGScrollWheelEventDeltaAxis1 = CGEventField.scrollWheelEventDeltaAxis1
private let kCGScrollWheelEventDeltaAxis2 = CGEventField.scrollWheelEventDeltaAxis2
private let kCGScrollWheelEventPointDeltaAxis1 = CGEventField.scrollWheelEventPointDeltaAxis1
private let kCGScrollWheelEventPointDeltaAxis2 = CGEventField.scrollWheelEventPointDeltaAxis2
private let kCGScrollWheelEventFixedPtDeltaAxis1 = CGEventField.scrollWheelEventFixedPtDeltaAxis1
private let kCGScrollWheelEventFixedPtDeltaAxis2 = CGEventField.scrollWheelEventFixedPtDeltaAxis2

@objc public enum MFDisplayLinkPhase: Int {
    case none = 0
    case start = 1
    case linear = 2
    case momentum = 3
    case end = 4
}

@objc(ScrollUtility)
public class ScrollUtility: NSObject {
    
    @objc public static let MFScrollPhaseToIOHIDEventPhase: NSDictionary = [
        MFDisplayLinkPhase.none.rawValue: kIOHIDEventPhaseUndefined,
        MFDisplayLinkPhase.start.rawValue: kIOHIDEventPhaseBegan,
        MFDisplayLinkPhase.linear.rawValue: kIOHIDEventPhaseChanged,
        MFDisplayLinkPhase.momentum.rawValue: kIOHIDEventPhaseChanged,
        MFDisplayLinkPhase.end.rawValue: kIOHIDEventPhaseEnded
    ]
    
    @objc public static func createPixelBasedScrollEvent(withValuesFrom event: CGEvent) -> CGEvent? {
        guard let newEvent = CGEvent(scrollWheelEvent2Source: nil, units: .pixel, wheelCount: 1, wheel1: 0, wheel2: 0, wheel3: 0) else {
            return nil
        }
        let valueFields = [
            kCGScrollWheelEventDeltaAxis1,
            kCGScrollWheelEventDeltaAxis2,
            kCGScrollWheelEventPointDeltaAxis1,
            kCGScrollWheelEventPointDeltaAxis2,
            kCGScrollWheelEventFixedPtDeltaAxis1,
            kCGScrollWheelEventFixedPtDeltaAxis2
        ]
        for field in valueFields {
            let ogVal = event.getIntegerValueField(field)
            newEvent.setIntegerValueField(field, value: ogVal)
        }
        return newEvent
    }
    
    @objc public static func createNormalizedEvent(withPixelValue lineHeight: Int32) -> CGEvent? {
        guard let event = CGEvent(scrollWheelEvent2Source: nil, units: .pixel, wheelCount: 1, wheel1: 0, wheel2: 0, wheel3: 0) else {
            return nil
        }
        event.setIntegerValueField(kCGScrollWheelEventDeltaAxis1, value: 1)
        event.setIntegerValueField(kCGScrollWheelEventPointDeltaAxis1, value: Int64(lineHeight))
        event.setDoubleValueField(kCGScrollWheelEventFixedPtDeltaAxis1, value: Double(lineHeight))
        
        DDLogInfo("Normalized scroll event values:")
        DDLogInfo("\(event.getIntegerValueField(kCGScrollWheelEventDeltaAxis1))")
        DDLogInfo("\(event.getIntegerValueField(kCGScrollWheelEventPointDeltaAxis1))")
        DDLogInfo("\(event.getDoubleValueField(kCGScrollWheelEventFixedPtDeltaAxis1))")
        
        return event
    }
    
    @objc public static func invertScrollEvent(_ event: CGEvent, direction: Int32) -> CGEvent {
        let dir = Int64(direction)
        // invert vertical
        let line1 = event.getIntegerValueField(kCGScrollWheelEventDeltaAxis1)
        let point1 = event.getIntegerValueField(kCGScrollWheelEventPointDeltaAxis1)
        let fixedPt1 = event.getIntegerValueField(kCGScrollWheelEventFixedPtDeltaAxis1)
        event.setIntegerValueField(kCGScrollWheelEventDeltaAxis1, value: line1 * dir)
        event.setIntegerValueField(kCGScrollWheelEventPointDeltaAxis1, value: point1 * dir)
        event.setIntegerValueField(kCGScrollWheelEventFixedPtDeltaAxis1, value: fixedPt1 * dir)
        // invert horizontal
        let line2 = event.getIntegerValueField(kCGScrollWheelEventDeltaAxis2)
        let point2 = event.getIntegerValueField(kCGScrollWheelEventPointDeltaAxis2)
        let fixedPt2 = event.getIntegerValueField(kCGScrollWheelEventFixedPtDeltaAxis2)
        event.setIntegerValueField(kCGScrollWheelEventDeltaAxis2, value: line2 * dir)
        event.setIntegerValueField(kCGScrollWheelEventPointDeltaAxis2, value: point2 * dir)
        event.setIntegerValueField(kCGScrollWheelEventFixedPtDeltaAxis2, value: fixedPt2 * dir)
        return event
    }
    
    @objc public static func makeScrollEventHorizontal(_ event: CGEvent) -> CGEvent {
        let line1 = event.getIntegerValueField(kCGScrollWheelEventDeltaAxis1)
        let point1 = event.getIntegerValueField(kCGScrollWheelEventPointDeltaAxis1)
        let fixedPt1 = event.getIntegerValueField(kCGScrollWheelEventFixedPtDeltaAxis1)
        
        event.setIntegerValueField(kCGScrollWheelEventDeltaAxis1, value: 0)
        event.setIntegerValueField(kCGScrollWheelEventPointDeltaAxis1, value: 0)
        event.setIntegerValueField(kCGScrollWheelEventFixedPtDeltaAxis1, value: 0)
        
        event.setIntegerValueField(kCGScrollWheelEventDeltaAxis2, value: line1)
        event.setIntegerValueField(kCGScrollWheelEventPointDeltaAxis2, value: point1)
        event.setIntegerValueField(kCGScrollWheelEventFixedPtDeltaAxis2, value: fixedPt1)
        
        return event
    }
    
    @objc public static func logScrollEvent(_ event: CGEvent) {
        let line1 = event.getIntegerValueField(kCGScrollWheelEventDeltaAxis1)
        let point1 = event.getIntegerValueField(kCGScrollWheelEventPointDeltaAxis1)
        let fixedPt1 = event.getIntegerValueField(kCGScrollWheelEventFixedPtDeltaAxis1)
        
        let line2 = event.getIntegerValueField(kCGScrollWheelEventDeltaAxis2)
        let point2 = event.getIntegerValueField(kCGScrollWheelEventPointDeltaAxis2)
        let fixedPt2 = event.getIntegerValueField(kCGScrollWheelEventFixedPtDeltaAxis2)
        
        DDLogInfo("Axis 1:")
        DDLogInfo("  Line: \(line1)")
        DDLogInfo("  Point: \(point1)")
        DDLogInfo("  FixedPt: \(fixedPt1)")
        
        DDLogInfo("Axis 2:")
        DDLogInfo("  Line: \(line2)")
        DDLogInfo("  Point: \(point2)")
        DDLogInfo("  FixedPt: \(fixedPt2)")
    }
    
    @objc public static func point(_ p1: CGPoint, isAboutTheSameAs p2: CGPoint, threshold: Int32) -> Bool {
        if abs(Int(p2.x - p1.x)) > Int(threshold) || abs(Int(p2.y - p1.y)) > Int(threshold) {
            return false
        }
        return true
    }
    
    @objc public static func sameSign(_ n: Double, and m: Double) -> Bool {
        if n == 0 || m == 0 {
            return true
        }
        if mfsign(n) == mfsign(m) {
            return true
        }
        return false
    }
    
    @objc public static func axisForVerticalDelta(_ deltaV: Int64, horizontalDelta deltaH: Int64) -> MFAxis {
        assert(deltaV == 0 || deltaH == 0, "Scroll event is not parallel to an axis.")
        var axis = kMFAxisVertical
        if deltaH != 0 {
            axis = kMFAxisHorizontal
        }
        return axis
    }
    
    @objc public static func directionForInputAxis(_ axis: MFAxis,
                                                  inputDelta: Int64,
                                                  invertSetting: MFScrollInversion,
                                                  horizontalModifier: Bool) -> MFDirection {
        let inputAxisDirection = mfsign(Double(inputDelta))
        let effectiveAxisDirection = Int32(inputAxisDirection) * invertSetting.rawValue
        
        if axis == kMFAxisHorizontal || horizontalModifier {
            if effectiveAxisDirection == -1 {
                return kMFDirectionLeft
            } else {
                return kMFDirectionRight
            }
        } else {
            if effectiveAxisDirection == -1 {
                return kMFDirectionDown
            } else {
                return kMFDirectionUp
            }
        }
    }
    
    private static var _mouseDidMove: Bool = false
    @objc public static var mouseDidMove: Bool {
        return _mouseDidMove
    }
    
    private static var _previousMouseLocation: CGPoint = .zero
    @objc public static func updateMouseDidMoveWithEvent(_ event: CGEvent) {
        let mouseLocation = event.location
        _mouseDidMove = !ScrollUtility.point(mouseLocation,
                                             isAboutTheSameAs: _previousMouseLocation,
                                             threshold: 10)
        _previousMouseLocation = mouseLocation
    }
    
    private static var _frontMostAppDidChange: Bool = false
    @objc public static var frontMostAppDidChange: Bool {
        return _frontMostAppDidChange
    }
    
    private static var _previousFrontMostApp: NSRunningApplication?
    @objc public static func updateFrontMostAppDidChange() {
        let frontMostApp = NSWorkspace.shared.frontmostApplication
        _frontMostAppDidChange = (frontMostApp != _previousFrontMostApp)
        _previousFrontMostApp = frontMostApp
    }
}

#endif
